stores/metrics.go

```diff
@@ -348,9 +389,42 @@ func (s *SQLStore) contractMetrics(ctx context.Context, start time.Time, n uint6
 	}
 
 	var metrics []dbContractMetric
-	err := s.findPeriods(tx, &metrics, start, n, interval)
-	if err != nil {
-		return nil, fmt.Errorf("failed to fetch contract metrics: %w", err)
+	if opts.ContractID == (types.FileContractID{}) && opts.HostKey == (types.PublicKey{}) {
+		// if neither contract nor host filters were set, we return the
+		// aggregate spending for each period
+		var metricsWithPeriod []struct {
+			Metric dbContractMetric `gorm:"embedded"`
+			Period int64
+		}
+		currentPeriod := int64(math.MinInt64)
+		err := tx.Raw(`
+			SELECT * FROM contracts
+			INNER JOIN (
+			SELECT fcid, MIN(timestamp) as timestamp, ? AS Period
+			FROM contracts
+			GROUP BY Period, fcid) i
+			ON contracts.fcid = i.fcid AND contracts.timestamp = i.timestamp
+		`, roundPeriodExpr(tx, start, interval)).
+			Scan(&metricsWithPeriod).
+			Error
+		if err != nil {
+			return nil, fmt.Errorf("failed to fetch aggregate metrics: %w", err)
+		}
+		for _, m := range metricsWithPeriod {
+			if m.Period != currentPeriod {
+				metrics = append(metrics, m.Metric)
+				currentPeriod = m.Period
+			} else {
+				metrics[len(metrics)-1] = metrics[len(metrics)-1].Aggregate(m.Metric)
```

Comment:
I'm not too sure but often with logic like this you have to perform this block outside of the loop to cover the last batch, is this one of those cases? e.g. do you need to call `Aggregate` on the last metric here after the loop on `metricsWithPeriod`?

Reply:
I think this is fine since every metric in the loop is either appended or aggregated. Looking at the test I also don't think there is anything missing at the end.